### PR TITLE
Rework Webmachine error logging to use the built-in log handler

### PR DESCRIPTION
**Pull request by kellymclaughlin**
Rework error logging to be more flexible. This changes the default error logging behavior from using the Erlang `error_logger` to using an event-based system that allows for more customization and flexibility for users. The default error logging handler provided (`webmachine_error_log_handler`) writes output directly to a file, but it is straightforward to use this module as a template and create a log handler to output via `error_logger` or `lager` or any other means. 

These changes bring error logging in line with how access and performance logging currently work. An advantage of this approach is that it allows error logging to be dynamically added and removed from applications using webmachine. It also allows for multiple error log handlers to be used simultaneously. Finally, it provides the benefit of not explicitly tying error logging  to a single mechanism that may not be the best choice for all applications using webmachine.

This PR changes the name of the access log handler module from `webmachine_log_handler` to the more descriptive `webmachine_access_log_handler`. It also updates the demo application to use the new event-based logging in its initialization.

### CODE REVIEW

**Comment by andrewjstone** [comment]
Running `make test` results in the following output: 
=ERROR REPORT==== 15-Aug-2013::13:34:40 ===
*\* Generic server webmachine_logger_watcher_sup terminating
*\* Last message in was {'EXIT',<0.184.0>,killed}
*\* When Server state == {state,
                            {local,webmachine_logger_watcher_sup},
                            simple_one_for_one,
                            [{child,undefined,webmachine_logger_watcher,
                                 {webmachine_logger_watcher,start_link,[]},
                                 transient,5000,worker,
                                 [webmachine_logger_watcher]}],
                            undefined,10,60,[],webmachine_logger_watcher_sup,
                            []}
*\* Reason for termination ==
*\* killed
....................................................................................................
OK, passed 100 tests
  All 60 tests passed.
Cover analysis: /Users/ajs/webmachine/.eunit/index.html

File: demo/src/webmachine_demo.erl [hunk 1]
```diff
@@ -16,7 +16,7 @@ ensure_started(App) ->
 start_link() ->
     ensure_started(crypto),
     ensure_started(mochiweb),
-    application:set_env(webmachine, webmachine_logger_module, 
+    application:set_env(webmachine, webmachine_logger_module,
```
Comment (andrewjstone) [inline demo/src/webmachine_demo.erl]: This is probably a misunderstanding on my part, but why set the `webmachine_logger_module` in `start_link/0` and `log_handlers` in `start/0`

File: src/webmachine_log.erl [hunk 2]
```diff
@@ -131,9 +135,24 @@ log_access(#wm_log_data{}=LogData) ->
...
+log_error(LogMsg) ->
+    gen_event:sync_notify(?EVENT_LOGGER, {log_error, LogMsg}).
+
+%% @doc Notify registered log event handler of an error event.
+-spec log_error(pos_integer(), #wm_reqdata{}, term()) -> ok.
+log_error(Code, Req, Reason) ->
+    gen_event:sync_notify(?EVENT_LOGGER, {log_error, Code, Req, Reason}).
+
+%% @doc Notify registered log event handler of an info event.
+-spec log_info(list()) -> ok.
```
Comment (andrewjstone) [inline src/webmachine_log.erl]: should prob be an iolist()

File: src/webmachine_log.erl [hunk 3]
```diff
@@ -131,9 +135,24 @@ log_access(#wm_log_data{}=LogData) ->
 %% @doc Close a log file.
 -spec log_close(atom(), string(), file:io_device()) -> ok | {error, term()}.
 log_close(Mod, Name, FD) ->
-    error_logger:info_msg("~p: closing log file: ~p~n", [Mod, Name]),
+    log_info([Mod, ": closing log file: ", Name, $\n]),
     file:close(FD).
 
+%% @doc Notify registered log event handler of an error event.
+-spec log_error(list()) -> ok.
```
Comment (andrewjstone) [inline src/webmachine_log.erl]: iolist() ?

**Comment by andrewjstone** [comment]
It's not part of the PR, but I noticed that [here](https://github.com/basho/webmachine/blob/080e4352495954510df7cd7bf960bcffd0ea5724/src/webmachine_log.erl#L176) it appears the iolist is unnecessarily being flattened. We may want to change that for performance reasons.

**Comment by andrewjstone** [comment]
@kellymclaughlin Code looks good. I've never reviewed a webmachine PR before though. Is there anything besides code review and running the tests that I should try?

**Comment by kellymclaughlin** [comment]
> Running make test results in the following output: 
> =ERROR REPORT==== 15-Aug-2013::13:34:40 ===
> *\* Generic server webmachine_logger_watcher_sup terminating
> *\* Last message in was {'EXIT',,killed}
> *\* When Server state == {state,
> {local,webmachine_logger_watcher_sup},
> simple_one_for_one,
> [{child,undefined,webmachine_logger_watcher,
> {webmachine_logger_watcher,start_link,[]},
> transient,5000,worker,
> [webmachine_logger_watcher]}],
> undefined,10,60,[],webmachine_logger_watcher_sup,
> []}
> *\* Reason for termination ==
> *\* killed
> ....................................................................................................
> OK, passed 100 tests
> All 60 tests passed.
> Cover analysis: /Users/ajs/webmachine/.eunit/index.html

This is just noisy shutdown. I have a fix for it, but it will be in a different PR.

**Comment by kellymclaughlin** [comment]
>  Code looks good. I've never reviewed a webmachine PR before though. Is there anything besides code review and running the tests that I should try?

I generally run the Riak CS `riak_test` suite using the PR branch.

File: demo/src/webmachine_demo.erl (hunk 1 above)
Comment (kellymclaughlin) [inline demo/src/webmachine_demo.erl]: Should probably have it in both, I'll add that.

**Comment by kellymclaughlin** [comment]
> It's not part of the PR, but I noticed that here it appears the iolist is unnecessarily being flattened. We may want to change that for performance reasons.

Addressed by [8ce5cc6](https://github.com/basho/webmachine/commit/8ce5cc6accaef6dd4c46e4a2be48321dc40b58a9)

**Comment by andrewjstone** [comment]
Riak Tests Pass :) :+1:

**Comment by andrewjstone** [comment]
:dancer:

**Pull request merged**